feature: restoreTransactionsProgrammatically

aboedo opened the pull request:
Added a silent version of `restoreTransactions`, called `restoreTransactionsProgrammatically`. It will be useful for developers performing migrations for other systems.

The new method is exactly the same as `restoreTransactions`, but without refreshing the receipt. 

The tests are also exactly like the ones for `restoreTransactions`, except for the one that asserts that the receipt isn't refreshed. 

I added a new enum `RCRefreshReceiptPolicy`, which specifies what the policy should be for receipt refresh:
 - always refresh
- never refresh
- refresh only if the receipt is empty, which we do for most stuff

aboedo commented on Purchases/Public/RCPurchases.h:
```diff
@@ -288,11 +288,23 @@ NS_SWIFT_NAME(purchasePackage(_:_:));
+ the same `appUserId` used to purchase originally.
+
+ @note This method will not trigger a login prompt from App Store. However, if the receipt currently on the device does not contain subscriptions, but the user has made subscription purchases, this method won't be able to restore them. Use restoreTransactionsWithCompletionBlock to cover those cases.
+ */
```
@rkotzy please take a look at the copy when you get the chance

MiguelCarranza commented on Purchases/Public/RCPurchases.h:
```diff
@@ -283,16 +283,34 @@ NS_SWIFT_NAME(purchaseProduct(_:_:));
+ the same `appUserId` used to purchase originally.
+
+ @note This method will not trigger a login prompt from App Store. However, if the receipt currently on the device does not contain subscriptions,
+ but the user has made subscription purchases, this method won't be able to restore them. Use restoreTransactionsWithCompletionBlock to cover those cases.
```
👍

vegaro commented on Purchases/Public/RCPurchases.h:
```diff
@@ -283,16 +283,34 @@ NS_SWIFT_NAME(purchaseProduct(_:_:));
+ You shouldn't use this method if you have your own account system. In that case "restoration" is provided by your app passing
+ the same `appUserId` used to purchase originally.
+
+ @note This method will not trigger a login prompt from App Store. However, if the receipt currently on the device does not contain subscriptions,
```
If the receipt is nil, it will trigger a login prompt since it will refresh it right?

vegaro replied:
I think that was the previous behavior before we forced to refresh at all times

aboedo replied:
ohh good catch. if that's the case, maybe it's a good idea to modify this a bit more and ensure that we're _never_ prompting a restore on the silent version

aboedo replied:
I made some updates to make sure that the silent version **never** refreshes

aboedo commented on Purchases/Public/RCPurchases.h:
```diff
@@ -283,16 +283,35 @@ NS_SWIFT_NAME(purchaseProduct(_:_:));
+ @note This method will not trigger a login prompt from App Store. However, if the receipt currently on the device does not contain subscriptions,
+ but the user has made subscription purchases, this method won't be able to restore them. Use restoreTransactionsWithCompletionBlock to cover those cases.
+ */
+- (void)restoreTransactionsProgrammaticallyWithCompletionBlock:(nullable RCReceivePurchaserInfoBlock)completion
+NS_SWIFT_NAME(restoreTransactionsProgrammatically(_:));
```
renamed after chatting with Jacob

aboedo commented on Purchases/Public/RCPurchases.m:
```diff
@@ -573,17 +574,27 @@ - (void)        purchaseProduct:(SKProduct *)product
 }
 
+- (void)restoreTransactionsProgrammaticallyWithCompletionBlock:(nullable RCReceivePurchaserInfoBlock)completion {
+    [self restoreTransactionsWithReceiptRefreshPolicy:RCReceiptRefreshPolicyNever completionBlock:completion];
```
moved from a boolean to this enum, I think it makes the intent pretty clear

vegaro replied:
I love it

aboedo commented on PurchasesTests/Purchasing/PurchasesTests.swift:
```diff
@@ -1059,6 +1062,140 @@ class PurchasesTests: XCTestCase {
+        expect(self.requestFetcher.refreshReceiptCalled) == false
+    }
+
+    func testRestoringPurchasesProgrammaticallyDoesntRefreshTheReceiptIfEmpty() {
+        setupPurchases()
+        self.receiptFetcher.shouldReturnReceipt = false
+        purchases!.restoreTransactionsProgrammatically()
+
+        expect(self.receiptFetcher.receiptDataTimesCalled) == 1
+        expect(self.requestFetcher.refreshReceiptCalled) == false
```
@vegaro this test would have failed before the changes with the enum

aboedo commented:
@vegaro @jeiting @MiguelCarranza 
changes since the last review: 
- moved from a boolean `shouldRefresh` to an enum `refreshPolicy`: `always | onlyWhenEmpty | never`
- renamed `silentRestoreTransactions` -> `restoreTransactionsProgrammatically`

vegaro commented on Purchases/Public/RCPurchases.m:
```diff
@@ -978,17 +989,20 @@ - (void)handleOfferingsUpdateError:(NSError *)error completion:(RCReceiveOfferin
     NSData *receiptData = [self.receiptFetcher receiptData];
-    if (receiptData == nil || receiptData.length == 0) {
+    BOOL receiptIsEmpty = receiptData == nil || receiptData.length == 0;
+    if (receiptIsEmpty && refreshPolicy == RCReceiptRefreshPolicyOnlyIfEmpty) {
```
I just realized that in `paymentDiscountForProductDiscount` we are not checking if the receipt is empty, we are just doing:

```
 [self receiptData:^(NSData *data) {
        [self.backend postOfferForSigning:discount.identifier
```

So this change would possibly return an empty receipt to that function, and we should probably return `missingReceiptFileError`

aboedo replied:
@vegaro I'm confused, why wouldn't this change should keep the same behavior for `paymentDiscountForProductDiscount`?
It calls `receiptData`, which in turn calls `receiptDataWithReceiptRefreshPolicy:RCReceiptRefreshPolicyOnlyIfEmpty`, so this line would render the exact same result as before, right?

aboedo replied:
talked about it over slack. 
Since within `paymentDiscountForProductDiscount` we're always calling with `RCReceiptRefreshPolicyOnlyIfEmpty`, there's no change in behavior. 

However, Cesar did correctly spot that there's an incorrect assumption in the method where we assume that the receiptData will be non-null and non-empty. 
if the receipt data is nil, that would cause a crash: we'd try to parse nil into a base-64 string, which would no-op and return nil again, and that nil would be inserted into a dictionary, causing a crash because NSDictionaries can't have nil values. 

if it's empty, it would actually post to the backend, but with an empty string, which is an unnecessary backend call. And every time you call the backend without needing to, a panda dies. Don't kill pandas. 

I'm fixing this and adding a test case for it.

aboedo replied:
updated: https://github.com/RevenueCat/purchases-ios/pull/387/files/4a4186700f03dcb401e6fb8842fed2592472c872..ac8575073a1f87135b69a283896d0889b6f750e3
